internal/dot/graph.go

```diff
@@ -28,26 +50,34 @@ type Node struct {
 	Group    string
 }
 
-// Edge connects a node parameter to a node result.
-type Edge struct {
-	Param  *Node
-	Result *Node
+// Add updates the param and result nodes in the constructor and adds the constructor to the graph
+func (dg *Graph) Add(c *Ctor, params []*Node, results []*Node) {
+	c.Params = getNodes(dg, params)
+	c.Results = getNodes(dg, results)
+	dg.Ctors = append(dg.Ctors, c)
 }
 
-// Graph is the DOT-format graph in a Container.
-type Graph struct {
-	Edges []*Edge
+func (n *Node) str() string {
+	if n.Name != "" {
+		return fmt.Sprintf("%v[name=%q]", n.Type, n.Name)
+	} else if n.Group != "" {
+		return fmt.Sprintf("%v[group=%q]", n.Type, n.Group)
+	}
+
+	return n.Type
 }
 
-// Add adds the edges in node n into dg.
-func (dg *Graph) Add(params []*Node, results []*Node) {
-	edges := make([]*Edge, 0, len(params)*len(results))
+func getNodes(dg *Graph, nodes []*Node) []*Node {
+	ptrs := make([]*Node, len(nodes))
+	for i, node := range nodes {
+		k := node.str()
 
-	for _, param := range params {
-		for _, result := range results {
-			edges = append(edges, &Edge{Param: param, Result: result})
+		if n := dg.Nodes[k]; n != nil {
+			ptrs[i] = n
+		} else {
+			dg.Nodes[k] = node
```

Comment:
A more-detailed note on how this function behaves might be helpful here. 
Seeing as this is storing nodes into the `dg.Nodes` map, we might want to update the function name so that it better encapsulates this mutation. `getNodes` might lead to unexpected results (i.e. "A 'read' ended up mutating the 'Graph' unexpectedly")